src/test/java/com/palantir/gradle/versions/CheckOverbroadConstraintsTest.java

```diff
@@ -0,0 +1,322 @@
+                        .build());
+
+        return testCases.map(Arguments::of);
+    }
+
+    private void checkTestCaseInResources(TestCase testCase) throws IOException {
+        File baseDir = new File(basePath);
+        File[] allFiles = baseDir.listFiles();
+
+        if (allFiles == null) {
+            throw new RuntimeException("Base directory does not exist or is not a directory.");
+        }
+
+        Set<String> baseDirFileNames =
+                Arrays.stream(allFiles).map(File::getName).collect(Collectors.toSet());
+
+        assertThat(baseDirFileNames.contains(testCase.fileName))
+                .as("%s not in resource directory. Run tests locally to fix.", testCase.fileName)
+                .isTrue();
+
+        File resourceFile = Arrays.stream(allFiles)
+                .filter(file -> file.getName().equals(testCase.fileName))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("%s found in file names but could not retrieve the file.", testCase.fileName)));
+
+        String actualContent =
+                Files.readString(resourceFile.toPath(), StandardCharsets.UTF_8).trim();
+
+        assertThat(actualContent)
+                .as(
+                        "Contents of file '%s' do not match the expected test case. Run tests locally to fix.",
+                        testCase.fileName)
+                .isEqualTo(testCase.fileContents().trim());
+    }
+
+    private static void deleteOldTests(Path path) throws IOException {
+        if (!Files.exists(path)) {
+            return;
+        }
+        try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
+            for (Path entry : stream) {
+                if (Files.isDirectory(entry)) {
+                    deleteOldTests(entry);
+                    Files.delete(entry);
+                } else {
+                    Files.delete(entry);
+                }
+            }
+        }
+    }
+
+    public static final class TestCase {
+        private final String fileName;
+        private final VersionsProps versionsProps;
+        private final LockState lockState;
+        private final List<String> propsLines;
+        private final String lockLines;
+
+        private TestCase(Builder builder) {
+            this.fileName = builder.fileName;
+            this.versionsProps = builder.versionsProps;
+            this.lockState = builder.lockState;
+            this.propsLines = builder.propsLines;
+            this.lockLines = builder.lockLines;
+        }
+
+        public static Builder builder(String testName) {
+            return new Builder(testName);
+        }
+
+        public List<String> newPropsLines() {
+            Map<String, List<String>> oldToNewLines =
+                    CheckOverbroadConstraints.determineNewLines(versionsProps, lockState);
+            return CheckOverbroadConstraints.generateUpdatedPropsLines(oldToNewLines, propsLines);
```

Comment:
We should run the check overbroad constraints on the new versions.props just to verify it does actually fix the problem for each of the tests (don't want to have a case where the fix does not fully fix the issue and the check fails after applying it).